### ISSUE TITLE
Switch carousel loading attribute from lazy to eager for next images when carousel is in view

### ISSUE THREAD

**Issue by pjeweb** [alveusgg/alveusgg#226]
I think it would be good to more eagerly load additional (hidden) images in a carousel that gets visible in viewport (or when interacting?) to prevent load flickering when scrolling the carousel.

My idea would be to just use `IntersectionObserver` to check if the carousel is in the viewport and maybe limit the number of images to a number (maybe as a prop).

**Comment by nslee333** [comment]
Is this still current? I'll take a shot at it if it is.

Also - is it unnecessary for me to check everytime if an issue is current?

**Comment by pjeweb** [comment]
> Is this still current? I'll take a shot at it if it is.

This is still current imho. Unless @MattIPv4 thinks this is a bad idea. He also designed the carousel component.

One example is the picture carousel for stompy:
https://www.alveussanctuary.org/ambassadors/stompy

Another example, where i noticed this first, are the show and tell posts with more than four images:
https://www.alveussanctuary.org/show-and-tell

If you watch in the network tab of the developer tools for those carousels, the browser (at least chrome) will load the first three/four images on desktop once it is scrolled in view, but not the other images. Those will only load in once you start scrolling the carousel.

I would just say, try to keep the code fairly simple and composable so it does not inflate the code too much. The carousel is already a pretty complex component. Since the items are currently passed as ReactNodes into the carousel this might needs some thinking how to handle all of that.

Some options that I see:
- detect view intersection in carousel …
   - … provide intersection state via Context, maybe as a pair of hooks to encapsulate the Context  `useProvideParentIntersection(carouselRef)` and `const isCarouselVisible = useParentInteraction()`
   - … and pass state to items using a render prop, refactoring all uses of `<Carousel/>`
   - … and use and effect to change the loading attribute in the DOM (sth like `querySelectorAll('[loading="lazy"]')`)
- detect view intersection in the items …
   - … using an effect looking for the intersection of the parent (carousel with some attribute) using DOM apis (maybe `Element#closest()`), maybe using something global to deduplication the observers, maybe make it a hook à la `const isCarouselVisible = useClosestIntersection(itemRef, '[data-carousel]')`

> Also - is it unnecessary for me to check everytime if an issue is current?

I think it's good that you ask or at least announce that you want to work on something just in case!

**Comment by MattIPv4** [comment]
Rather than using a context/effect, I think I'd be most inclined to suggest a pattern similar to the YouTube Lightbox, where we allow children to be functions that can be passed a context (and similar to how many of the headless UI components operate): https://github.com/alveusgg/alveusgg/blob/ed84b25f2eb60061eefba5ac8b44f1678a1a0ea8/apps/website/src/components/content/YouTube.tsx#L232-L240

The carousel can then run intersection observers on each child in the carousel and keep track of what children are visible -- this can then be passed in that context and allow consumers that're passing children to the carousel to determine if an item is next up to become visible.

Some pseudo-code for the consumer that'd be using this, so you can get a sense of the interface I'm thinking of:

```jsx
const items = [
 <>item 1</>,
 <>item 2</>,
 <>item 3</>,
 <>item 4</>,
];

const mapped = items.map((item, idx) => (({ $visible }) => {
  // If the item before us is active, we're off-screen to the right (or on-screen)
  // If the item after us is active, we're off-screen to the left (or on-screen)
  const preload = $visible.includes(idx + 1) || $visible.includes(idx - 1);
  
  // Do whatever with preload
  return item;
}));

const page = () => (
  <>
    <Carousel items={mapped} />
    {/* whatever else ... */>
  </>
);
```

Thoughts?

**Comment by pjeweb** [comment]
> Thoughts?

Looks great. Thats pretty much what i meant with [render props](https://react.dev/reference/react/cloneElement#passing-data-with-a-render-prop), but your solution would also allow passing simple nodes, not only render props.

**Comment by nslee333** [comment]
Alright I'm working on learning the carousel and getting ready to open a draft PR.

I had to delete my fork because I was committing to main and messed up the commit history and now trying to get my dev environment working again. 

Have you ran into this issue before? I reran through the `local development` section multiple times. I must be doing something wrong. 

`./src/assets/socials/georgie-reel-clip.mp4?quality=low
NonErrorEmittedError: (Emitted value instead of an instance of Error) [object Object]`

`index.js:619 Uncaught ModuleBuildError: Module build failed (from ./build-scripts/video-loader.cjs):
NonErrorEmittedError: (Emitted value instead of an instance of Error) [object Object]`

**Comment by MattIPv4** [comment]
Any other logs prior to that point? Not an error I've seen before with the loader

**Comment by nslee333** [comment]
Processing video /home/nslee333/cs/alveusgg/apps/website/src/assets/socials/georgie-reel-clip.mp4 ...
│  ... /home/nslee333/cs/alveusgg/apps/website/src/assets/socials/georgie-reel-clip.mp4 failed {
│   error: Error: Command failed: "/home/nslee333/cs/alveusgg/node_modules/.pnpm/@ffprobe-installer+ffprobe@2.0.0/node_modules/@ffprobe-installer/linux-x64/ffprobe" -v…
│   /bin/sh: 1: /home/nslee333/cs/alveusgg/node_modules/.pnpm/@ffprobe-installer+ffprobe@2.0.0/node_modules/@ffprobe-installer/linux-x64/ffprobe: Permission denied
│
│       at ChildProcess.exithandler (node:child_process:419:12)
│       at ChildProcess.emit (node:events:511:28)
│       at maybeClose (node:internal/child_process:1098:16)
│       at Socket.<anonymous> (node:internal/child_process:456:11)
│       at Socket.emit (node:events:511:28)
│       at Pipe.<anonymous> (node:net:334:12) {
│     code: 126,
│     killed: false,
│     signal: null,
│     cmd: '"/home/nslee333/cs/alveusgg/node_modules/.pnpm/@ffprobe-installer+ffprobe@2.0.0/node_modules/@ffprobe-installer/linux-x64/ffprobe" -v error -select_streams…
│   },
│   stdout: '',
│   stderr: '/bin/sh: 1: /home/nslee333/cs/alveusgg/node_modules/.pnpm/@ffprobe-installer+ffprobe@2.0.0/node_modules/@ffprobe-installer/linux-x64/ffprobe: Permission d…
│ }
│ - error ./src/assets/socials/georgie-reel-clip.mp4?quality=low
│ NonErrorEmittedError: (Emitted value instead of an instance of Error) [object Object]
│ Import trace for requested module:
│ ./src/assets/socials/georgie-reel-clip.mp4?quality=low
│ ./src/components/layout/footer/Socials.tsx
│ ./src/components/layout/footer/Footer.tsx
│ ./src/components/layout/Layout.tsx

I got this log repeated 4 times, I saw the ffprobe permission denied bit.
Tried deleting node_modules folders and pushed prisma db, No change
Tried restarting machine. No change.

I'm seeing a few stack overflow posts related to ffprobe and ffmpeg permission denied might try those solutions.

**Comment by MattIPv4** [comment]
How very odd, yeah I've never seen that before. Almost feels like pnpm is breaking permissions somehow when doing the install 🤔 If you don't get any further with it, please open a dedicated issue for it (it's rather unrelated to what this issue is discussing).

**Comment by nslee333** [comment]
>The carousel can then run intersection observers on each child in the carousel and keep track of what children are visible -- this can then be passed in that context and allow consumers that're passing children to the carousel to determine if an item is next up to become visible.

In summary: 
The context (which item is visible) is lifted up from the carousel to the consumer (parent), then the consumer preloads the next item to be viewed?

Trying to wrap my head around this

**Comment by MattIPv4** [comment]
Yes, that's correct. The Carousel tracks what is visible, and exposes that as a context to its children, which are provided by the parent. The parent can then use that context in the children it passes to the Carousel to preload items etc.

**Comment by nslee333** [comment]
Sorry guys I'm going to bow out of this one, don't have time to help anymore.

**Comment by MattIPv4** [comment]
No worries, greatly appreciate the work you put into it, and the PR will be a good reference for whoever tackles this next 💚
